### PR TITLE
Parseable concepts to simplify SFINAE

### PR DESCRIPTION
**Pull request by JoeLoser**
### :notebook_with_decorative_cover: Description

Problem:
- There is a lot of variable templates with cumbersome metaprogramming
  in defining them.
- There is some verbose SFINAE for constraining various parseable
  concepts.

Solution:
- Add various concepts and apply them to simplify the metaprogramming.

### :memo: Checklist

- [X] All user-facing changes have changelog entries.
- [X] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [X] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

Commit-by-commit.

### CODE REVIEW

File: libvast/vast/access.hpp
Comment (mavam) [inline libvast/vast/access.hpp:15]: Something for the future: this type exists only to give some parsers/printers access to the private state of befriended types. It's redundant to our `inspect` overloads, which we should use instead. `git grep` reveals only `address`, `subnet`, and `pattern` still using this.

File: libvast/vast/concept/parseable/core/parser.hpp
Comment (mavam) [inline libvast/vast/concept/parseable/core/parser.hpp:122]:
```suggestion
concept parser = is_parser<std::decay_t<T>>::value;
```

Would this be possible, or be name conflict? (Same for `printer`.)

File: libvast/vast/concept/parseable/to.hpp
Comment (mavam) [inline libvast/vast/concept/parseable/to.hpp:30]: `Range` would also make for a great concept, with a specialization for character ranges for this function.

**Review by mavam — COMMENTED** [review]
All great stuff, thanks!

I wonder how we deal with naming conflicts, e.g., the `parseable` concept vs. the `parser` base class. I found [P1754](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1754r0.pdf) on this, which suggests the following:
- Use a name that is an adjective or abstract noun. (Prefer names like `regular` and `swappable` and `color`. Avoid names like `has constructor` and `red.`)
- Occasionally, use a `_type` suffix as a concession to avoiding name collisions, usually for very general concepts with very common names.

The reason why I'm bringing this up is that `is_parser_v` doesn't really look like a concept. I understand that it makes for a drop-in replacement of the type trait, but as we gain experience with concepts, this is a good discussion opportunity.

File: libvast/vast/concept/parseable/core/parser.hpp
Comment (JoeLoser) [reply libvast/vast/concept/parseable/core/parser.hpp:122]: Good thought -- it would be a name conflict. I was planning on renaming some things in a follow-up PR. E.g. change the `parser` class type to be `parser_t` and then rename the concept `is_parser_v` to `parser`. Same with `printer`. I thought that would introduce more noise than value for review purposes of this PR.

File: libvast/vast/access.hpp
Comment (JoeLoser) [reply libvast/vast/access.hpp:15]: Thanks for the insight. We should clean that up in a follow-up PR then.

File: libvast/vast/concept/parseable/to.hpp
Comment (JoeLoser) [reply libvast/vast/concept/parseable/to.hpp:30]: Agreed. I'm starting to see more places where good concepts could help as I'm perusing more of the codebase with my C++20 lens in mind.

File: libvast/vast/concept/parseable/core/parser.hpp
Comment (mavam) [reply libvast/vast/concept/parseable/core/parser.hpp:122]: Nice idea. I agree that the renaming could be a bit noisy in here. Although in general, with clear review instructions, you can make the reviewer look at your PR the way you want...most of the time. 😉 

Regarding the concrete renaming: how about we use something like `basic_parser` or `parser_base`? It's a bit more semantic that `*_t`.

**Review by mavam — APPROVED** [review]
We discussed the follow-up renaming in the conversation. With this being part 1 out of 2, green light from my side.

(I'll let @tenzir/backend merge, though, in case I'm missing something obvious with concepts.)

File: libvast/vast/concept/parseable/core/parser.hpp
Comment (JoeLoser) [reply libvast/vast/concept/parseable/core/parser.hpp:122]: I don't feel strongly on the rename. I like `parser_base` too, so we can go with that.

**Comment by lava** [comment]
Hi, I'm adding a 'blocked'-label since we're currently in feature freeze until our next release on thursday.

Based on a quick glance this looks good though, so if you want to continue conceptifying things feel free to continue off of this branch until then :D

**Comment by JoeLoser** [comment]
> Hi, I'm adding a 'blocked'-label since we're currently in feature freeze until our next release on thursday.
> 
> Based on a quick glance this looks good though, so if you want to continue conceptifying things feel free to continue off of this branch until then :D

Sounds good, thanks for the heads up 👍

**Comment by JoeLoser** [comment]
> We discussed the follow-up renaming in the conversation. With this being part 1 out of 2, green light from my side.
> 
> (I'll let @tenzir/backend merge, though, in case I'm missing something obvious with concepts.)

I just pushed a few commits to do some renaming of the concepts and `parser -> parser_base` to avoid the name conflict.

**Review by mavam — APPROVED** [review]
Looked at the new changes and they work out nicely. Looking forward to merge this after Thu!

**Pull request merged**
